File: src/main/java/me/anonymoussoftware/vacancymanager/ui/panels/EmployerListPanel.java
package me.anonymoussoftware.vacancymanager.ui.panels;

import java.awt.Color;
import java.awt.Component;
import java.awt.EventQueue;
import java.util.List;

import javax.swing.BoxLayout;
import javax.swing.DefaultListModel;
import javax.swing.JLabel;
import javax.swing.JList;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.ListCellRenderer;
import javax.swing.ListSelectionModel;

import me.anonymoussoftware.vacancymanager.App;
import me.anonymoussoftware.vacancymanager.VacancyManager;
import me.anonymoussoftware.vacancymanager.VacancyManager.EmployerListChangeListener;
import me.anonymoussoftware.vacancymanager.model.Employer;

@SuppressWarnings("serial")
public class EmployerListPanel extends JPanel implements EmployerListChangeListener {

    private final DefaultListModel<Employer> employerListModel;

    private final VacancyManager vacancyManager = App.getBean(VacancyManager.class);

    private final JLabel label;

    private final JList<Employer> employerList;

    public EmployerListPanel() {
        setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));

        this.label = new JLabel("Companies:");
        add(this.label);

        this.employerListModel = new DefaultListModel<>();
        this.employerList = new JList<>(this.employerListModel);
        this.employerList.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
        this.employerList.setCellRenderer(new EmployerCellRenderer());
        JScrollPane vacancyListScrollPane = new JScrollPane(this.employerList);
        add(vacancyListScrollPane);
    }

    @Override
    public void addNotify() {
        super.addNotify();
        this.vacancyManager.addEmployerListChangeListener(this);
    }

    @Override
    public void removeNotify() {
        this.vacancyManager.removeEmployerListChangeListener(this);
        super.removeNotify();
    }

    @Override
    public void onEmployerListChange() {
        EventQueue.invokeLater(() -> {
            List<Employer> employers = this.vacancyManager.getEmployers();
            int selectedIndex = this.employerList.getSelectedIndex();
            this.employerListModel.clear();
            employers.stream().forEach(this.employerListModel::addElement);
            if (selectedIndex >= 0) {
                int employerListSize = this.employerList.getModel().getSize();
                this.employerList.setSelectedIndex(Math.min(selectedIndex, Math.max(employerListSize - 1, 0)));
            }
            this.label.setText("Companies: " + employers.size());
        });
    }

    private class EmployerCellRenderer extends JLabel implements ListCellRenderer<Employer> {

        @Override
        public Component getListCellRendererComponent(JList<? extends Employer> list, Employer value, int index,
                boolean isSelected, boolean cellHasFocus) {
            Employer selectedValue = list.getSelectedValue();
            Color backgroundColor;
            Color foregroundColor = list.getForeground();
            if (selectedValue != null && selectedValue.getId() == value.getId()) {
                backgroundColor = list.getSelectionBackground();
                foregroundColor = list.getSelectionForeground();
            } else if (value.isBanned()) {
                backgroundColor = Color.DARK_GRAY;
                foregroundColor = Color.WHITE;
            } else {
                backgroundColor = list.getBackground();
            }
            setBackground(backgroundColor);
            setForeground(foregroundColor);
            setOpaque(true);
            setFont(list.getFont());
            setText(value.getName());
            return this;
        }

    }

}

File: src/main/kotlin/me/anonymoussoftware/vacancymanager/ui/panels/EmployerListPanel.kt
package me.anonymoussoftware.vacancymanager.ui.panels

import me.anonymoussoftware.vacancymanager.App
import me.anonymoussoftware.vacancymanager.VacancyManager
import me.anonymoussoftware.vacancymanager.VacancyManager.EmployerListChangeListener
import me.anonymoussoftware.vacancymanager.model.Employer

import javax.swing.*
import java.awt.*

class EmployerListPanel : JPanel(), EmployerListChangeListener {

    private val employerListModel: DefaultListModel<Employer>

    private val vacancyManager = App.getBean(VacancyManager::class.java)

    private val label: JLabel

    private val employerList: JList<Employer>

    init {
        layout = BoxLayout(this, BoxLayout.Y_AXIS)

        this.label = JLabel("Companies:")
        add(this.label)

        this.employerListModel = DefaultListModel()
        this.employerList = JList(this.employerListModel)
        this.employerList.selectionMode = ListSelectionModel.SINGLE_SELECTION
        this.employerList.cellRenderer = EmployerCellRenderer()
        val vacancyListScrollPane = JScrollPane(this.employerList)
        add(vacancyListScrollPane)
    }

    override fun addNotify() {
        super.addNotify()
        this.vacancyManager.addEmployerListChangeListener(this)
    }

    override fun removeNotify() {
        this.vacancyManager.removeEmployerListChangeListener(this)
        super.removeNotify()
    }

    override fun onEmployerListChange() {
        EventQueue.invokeLater {
            val employers = this.vacancyManager.getEmployers()
            val selectedIndex = this.employerList.selectedIndex
            this.employerListModel.clear()
            employers.stream().forEach{ this.employerListModel.addElement(it) }
            if (selectedIndex >= 0) {
                val employerListSize = this.employerList.model.size
                this.employerList.selectedIndex = Math.min(selectedIndex, Math.max(employerListSize - 1, 0))
            }
            this.label.text = "Companies: " + employers.size
        }
    }

    private inner class EmployerCellRenderer : JLabel(), ListCellRenderer<Employer> {

        override fun getListCellRendererComponent(
            list: JList<out Employer>, value: Employer, index: Int,
            isSelected: Boolean, cellHasFocus: Boolean
        ): Component {
            val selectedValue = list.selectedValue
            val backgroundColor: Color
            var foregroundColor = list.foreground
            if (selectedValue != null && selectedValue.id == value.id) {
                backgroundColor = list.selectionBackground
                foregroundColor = list.selectionForeground
            } else if (value.isBanned) {
                backgroundColor = Color.DARK_GRAY
                foregroundColor = Color.WHITE
            } else {
                backgroundColor = list.background
            }
            background = backgroundColor
            foreground = foregroundColor
            isOpaque = true
            font = list.font
            text = value.name
            return this
        }

    }

}
